File: skills/product/design-brief/scripts/check-contrast.ts
#!/usr/bin/env bun
/**
 * check-contrast.ts — WCAG 2.x contrast checker for DESIGN.md color tokens.
 *
 * Execute this script; never read it as reference.
 *
 * Usage:
 *   bun run check-contrast.ts <path-to-DESIGN.md> [--json]
 *   bun run check-contrast.ts --pair <#RRGGBB> <#RRGGBB>
 *
 * File mode checks:
 *   - every `colors.<base>` / `colors.<base>-foreground` pair at 4.5:1
 *     (`foreground` itself pairs with `background`)
 *   - `colors.muted-foreground` against `colors.background` and
 *     `colors.card` at 4.5:1 (it doubles as secondary text there)
 *   - every `components.<name>` with both `backgroundColor` and
 *     `textColor` resolved, at 4.5:1 (3:1 noted for large text / UI)
 *   - `-disabled` component variants reported SKIP (inactive UI is
 *     exempt under WCAG 1.4.3)
 *
 * The `colors:` block may be flat, or grouped into skins (light/dark
 * or any other name). Skin groups are detected structurally — a child
 * map without a `hex`/`oklch` member is a group — never by name. Each
 * skin's pairs are checked independently; component references resolve
 * once per skin.
 *
 * Object-form colors are checked through their `hex` member, in both
 * block form (nested `hex:` line) and inline flow form
 * (`{ hex: "#...", oklch: "..." }`); hex↔oklch agreement is a separate
 * validate rule.
 *
 * A run where every pair skips verifies nothing and exits 2 — an
 * all-SKIP result must never read as a passing gate.
 *
 * Exit codes: 0 = no failures, 1 = at least one FAIL, 2 = usage/IO error.
 */

// Force module scope so the declarations below stay file-local instead
// of merging into the global scope shared with sibling scripts.
export {};

// Standalone script — no tsconfig or @types/node alongside it. Declaring
// the exact runtime members used here keeps editor diagnostics clean
// without a types dependency; exit() typed `never` restores TS flow
// narrowing after early exits.
declare const process: { argv: string[]; exit(code?: number): never };
declare function require(id: "fs"): {
  readFileSync(path: string, encoding: "utf8"): string;
};
const { readFileSync } = require("fs");

type Rgb = { r: number; g: number; b: number };

type Finding = {
  status: "PASS" | "FAIL" | "SKIP";
  ratio: number | null;
  pair: string;
  note?: string;
};

// Thresholds from WCAG 2.x SC 1.4.3 (AA): 4.5:1 for body text,
// 3:1 for large text and UI components.
const BODY_TEXT_RATIO = 4.5;
const LARGE_TEXT_RATIO = 3.0;

function unquote(value: string): string {
  return value.trim().replace(/^["']|["']$/g, "");
}

function parseHex(value: string): Rgb | null {
  const hex = unquote(value);
  const short = /^#([0-9a-f])([0-9a-f])([0-9a-f])$/i.exec(hex);
  if (short) {
    return {
      r: parseInt(short[1] + short[1], 16),
      g: parseInt(short[2] + short[2], 16),
      b: parseInt(short[3] + short[3], 16),
    };
  }
  const full = /^#([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})$/i.exec(hex);
  if (full) {
    return {
      r: parseInt(full[1], 16),
      g: parseInt(full[2], 16),
      b: parseInt(full[3], 16),
    };
  }
  return null;
}

// WCAG 2.x relative luminance: linearize each sRGB channel with the
// spec's piecewise curve (0.04045 threshold, 12.92 linear slope,
// 2.4 gamma), then weight by the Rec. 709 luma coefficients.
function relativeLuminance({ r, g, b }: Rgb): number {
  const linearize = (channel: number): number => {
    const c = channel / 255;
    return c <= 0.04045 ? c / 12.92 : Math.pow((c + 0.055) / 1.055, 2.4);
  };
  return 0.2126 * linearize(r) + 0.7152 * linearize(g) + 0.0722 * linearize(b);
}

// WCAG contrast ratio: (L_lighter + 0.05) / (L_darker + 0.05).
// The 0.05 term models ambient screen flare per the spec.
function contrastRatio(a: Rgb, b: Rgb): number {
  const la = relativeLuminance(a);
  const lb = relativeLuminance(b);
  return (Math.max(la, lb) + 0.05) / (Math.min(la, lb) + 0.05);
}

// Composite a translucent foreground over its background in sRGB space,
// matching how browsers blend non-color-managed content.
function blend(fg: Rgb, bg: Rgb, alpha: number): Rgb {
  const mix = (f: number, b: number) => Math.round(f * alpha + b * (1 - alpha));
  return { r: mix(fg.r, bg.r), g: mix(fg.g, bg.g), b: mix(fg.b, bg.b) };
}

function extractFrontmatter(markdown: string): string | null {
  const match = /^---\r?\n([\s\S]*?)\r?\n---/.exec(markdown);
  return match ? match[1] : null;
}

type ColorMap = Record<string, string>;
// Skin name -> token -> raw color value. The "" key is the default
// skin holding flat (ungrouped) tokens.
type Skins = Record<string, ColorMap>;

type YamlNode = string | { [key: string]: YamlNode };

// Minimal indentation walker: returns the subtree under a column-0
// `rootKey:` as nested maps of raw string values. Inline flow values
// (`{ ... }`) stay as the raw string for the caller to inspect.
function parseSubtree(yaml: string, rootKey: string): YamlNode | null {
  const lines = yaml.split(/\r?\n/);
  let start = -1;
  for (let i = 0; i < lines.length; i++) {
    if (lines[i] === `${rootKey}:` || lines[i].startsWith(`${rootKey}: `)) {
      start = i;
      break;
    }
  }
  if (start === -1) return null;
  const root: { [key: string]: YamlNode } = {};
  const stack: Array<{ indent: number; node: { [key: string]: YamlNode } }> = [
    { indent: 0, node: root },
  ];
  for (let i = start + 1; i < lines.length; i++) {
    const line = lines[i];
    if (!line.trim() || /^\s*#/.test(line)) continue;
    const entry = /^( +)([\w-]+):\s*(.*)$/.exec(line);
    if (!entry) break; // column-0 key or non-key line ends the block
    const indent = entry[1].length;
    while (stack.length > 1 && indent <= stack[stack.length - 1].indent) {
      stack.pop();
    }
    const parent = stack[stack.length - 1].node;
    if (entry[3]) {
      parent[entry[2]] = unquote(entry[3]);
    } else {
      const child: { [key: string]: YamlNode } = {};
      parent[entry[2]] = child;
      stack.push({ indent, node: child });
    }
  }
  return root;
}

// A color leaf carries its checkable value as a literal string, an
// inline flow map with a `hex:` member, or a block map with a `hex`
// key. Returns the raw candidate (parseHex validates later); a leaf
// with only `oklch` returns that string so the pair surfaces as SKIP
// instead of vanishing.
function colorLeafValue(node: YamlNode): string | null {
  if (typeof node === "string") {
    const flow = /^\{.*\bhex:\s*([^,}]+)/.exec(node);
    return flow ? unquote(flow[1]) : node;
  }
  if (typeof node.hex === "string") return unquote(node.hex);
  if (typeof node.oklch === "string") return unquote(node.oklch);
  return null;
}

function isColorLeaf(node: YamlNode): boolean {
  return typeof node === "string" || "hex" in node || "oklch" in node;
}

// Interpret the `colors:` subtree. Flat tokens land in the "" skin; a
// child map without a `hex`/`oklch` member is a skin group (light,
// dark, any name — groups are structural, never name-matched) holding
// its own token map.
function parseColors(yaml: string): Skins {
  const tree = parseSubtree(yaml, "colors");
  const skins: Skins = {};
  if (!tree || typeof tree === "string") return skins;
  const put = (skin: string, token: string, raw: string) => {
    (skins[skin] ??= {})[token] = raw;
  };
  for (const [key, value] of Object.entries(tree)) {
    if (isColorLeaf(value)) {
      const raw = colorLeafValue(value);
      if (raw !== null) put("", key, raw);
      continue;
    }
    for (const [token, tokenValue] of Object.entries(value)) {
      const raw = isColorLeaf(tokenValue) ? colorLeafValue(tokenValue) : null;
      if (raw !== null) put(key, token, raw);
    }
  }
  return skins;
}

// Minimal parser for `components:`: name -> { prop: rawValue }.
function parseComponents(yaml: string): Record<string, Record<string, string>> {
  const components: Record<string, Record<string, string>> = {};
  let inBlock = false;
  let current: string | null = null;
  for (const line of yaml.split(/\r?\n/)) {
    if (/^components:\s*$/.test(line)) {
      inBlock = true;
      continue;
    }
    if (inBlock && /^\S/.test(line)) break;
    if (!inBlock || !line.trim()) continue;
    const name = /^ {2}([\w-]+):\s*$/.exec(line);
    const prop = /^ {4}([\w-]+):\s*(\S.*)$/.exec(line);
    if (name) {
      current = name[1];
      components[current] = {};
    } else if (prop && current) {
      components[current][prop[1]] = unquote(prop[2]);
    }
  }
  return components;
}

// Resolve a component color prop: a literal hex value, or a
// `{colors.<token>}` reference with an optional `/NN` opacity
// modifier. References resolve in the active skin first, then the
// flat default skin; a dotted path (`{colors.<skin>.<token>}`) pins a
// specific skin.
function resolveColor(
  raw: string,
  colors: ColorMap,
  skins: Skins,
): { rgb: Rgb; alpha: number } | null {
  const ref = /^\{colors\.([\w.-]+)\}(?:\/(\d{1,3}))?$/.exec(unquote(raw));
  if (ref) {
    let hex: string | undefined = colors[ref[1]] ?? skins[""]?.[ref[1]];
    if (!hex && ref[1].includes(".")) {
      const dot = ref[1].indexOf(".");
      hex = skins[ref[1].slice(0, dot)]?.[ref[1].slice(dot + 1)];
    }
    if (!hex) return null;
    const rgb = parseHex(hex);
    if (!rgb) return null;
    const alpha = ref[2] ? Math.min(100, parseInt(ref[2], 10)) / 100 : 1;
    return { rgb, alpha };
  }
  const literal = parseHex(raw);
  return literal ? { rgb: literal, alpha: 1 } : null;
}

function formatRatio(ratio: number): string {
  return `${ratio.toFixed(2)}:1`;
}

function checkFile(path: string, asJson: boolean): never {
  let markdown: string;
  try {
    markdown = readFileSync(path, "utf8");
  } catch {
    console.error(`Cannot read ${path}`);
    process.exit(2);
  }
  const yaml = extractFrontmatter(markdown);
  if (!yaml) {
    console.error(`No YAML frontmatter found in ${path}`);
    process.exit(2);
  }
  const skins = parseColors(yaml);
  const skinNames = Object.keys(skins);
  if (skinNames.length === 0) {
    console.error(`No parseable color tokens found in ${path}`);
    process.exit(2);
  }
  const components = parseComponents(yaml);
  const findings: Finding[] = [];

  const check = (bg: Rgb, fg: Rgb, pair: string) => {
    const ratio = contrastRatio(bg, fg);
    findings.push({
      status: ratio >= BODY_TEXT_RATIO ? "PASS" : "FAIL",
      ratio,
      pair,
      note:
        ratio >= BODY_TEXT_RATIO
          ? undefined
          : ratio >= LARGE_TEXT_RATIO
            ? `needs ${BODY_TEXT_RATIO}:1 for body text; passes ${LARGE_TEXT_RATIO}:1 for large text and UI only`
            : `needs ${BODY_TEXT_RATIO}:1`,
    });
  };

  for (const skinName of skinNames) {
    const colors = skins[skinName];
    const label = (token: string) =>
      skinName ? `colors.${skinName}.${token}` : `colors.${token}`;

    const checkTokenPair = (baseKey: string, fgKey: string) => {
      const bg = parseHex(colors[baseKey]);
      const fg = parseHex(colors[fgKey]);
      if (!bg || !fg) {
        findings.push({
          status: "SKIP",
          ratio: null,
          pair: `${label(fgKey)} on ${label(baseKey)}`,
          note: "unparseable hex value",
        });
        return;
      }
      check(bg, fg, `${label(fgKey)} on ${label(baseKey)}`);
    };

    for (const name of Object.keys(colors)) {
      if (!name.endsWith("foreground")) continue;
      const base =
        name === "foreground" ? "background" : name.replace(/-foreground$/, "");
      if (base !== name && colors[base]) checkTokenPair(base, name);
    }

    // muted-foreground doubles as secondary text on background and card
    if (colors["muted-foreground"]) {
      for (const surface of ["background", "card"]) {
        if (colors[surface]) checkTokenPair(surface, "muted-foreground");
      }
    }
  }

  // Component references carry no skin name, so each pair resolves and
  // checks once per skin that can resolve it; a single SKIP surfaces
  // only when no skin resolves the pair.
  for (const [name, props] of Object.entries(components)) {
    const bgRaw = props.backgroundColor;
    const fgRaw = props.textColor;
    if (!bgRaw || !fgRaw) continue;
    if (/-disabled$/.test(name)) {
      findings.push({
        status: "SKIP",
        ratio: null,
        pair: `components.${name}`,
        note: "disabled state exempt (WCAG 1.4.3 excludes inactive UI)",
      });
      continue;
    }
    let resolvedAny = false;
    for (const skinName of skinNames) {
      const colors = skins[skinName];
      const bg = resolveColor(bgRaw, colors, skins);
      const fg = resolveColor(fgRaw, colors, skins);
      if (!bg || !fg) continue;
      resolvedAny = true;
      // A translucent background composites over the page; approximate the
      // page with colors.background, falling back to opaque white.
      const page = parseHex(
        colors.background ?? skins[""]?.background ?? "#FFFFFF",
      ) ?? { r: 255, g: 255, b: 255 };
      const bgRgb = bg.alpha < 1 ? blend(bg.rgb, page, bg.alpha) : bg.rgb;
      const fgRgb = fg.alpha < 1 ? blend(fg.rgb, bgRgb, fg.alpha) : fg.rgb;
      const suffix = skinName ? ` [${skinName}]` : "";
      check(
        bgRgb,
        fgRgb,
        `components.${name} textColor on backgroundColor${suffix}`,
      );
    }
    if (!resolvedAny) {
      findings.push({
        status: "SKIP",
        ratio: null,
        pair: `components.${name}`,
        note: "color did not resolve to a hex value",
      });
    }
  }

  const fails = findings.filter((f) => f.status === "FAIL").length;
  const skips = findings.filter((f) => f.status === "SKIP").length;
  if (asJson) {
    console.log(JSON.stringify(findings, null, 2));
  } else {
    for (const f of findings) {
      const ratio = f.ratio === null ? "  --  " : formatRatio(f.ratio).padStart(7);
      console.log(`${f.status.padEnd(4)} ${ratio}  ${f.pair}${f.note ? ` (${f.note})` : ""}`);
    }
    console.log(`\n${findings.length} pairs checked, ${fails} failing, ${skips} skipped`);
  }
  // An all-SKIP run verified nothing; never let it read as a pass.
  if (findings.length > 0 && skips === findings.length) {
    console.error(
      "Every pair was skipped — nothing verified. Check the color token shapes.",
    );
    process.exit(2);
  }
  process.exit(fails > 0 ? 1 : 0);
}

function checkPair(a: string, b: string): never {
  const colorA = parseHex(a);
  const colorB = parseHex(b);
  if (!colorA || !colorB) {
    console.error('Usage: check-contrast.ts --pair "#RRGGBB" "#RRGGBB"');
    process.exit(2);
  }
  const ratio = contrastRatio(colorA, colorB);
  const body = ratio >= BODY_TEXT_RATIO ? "PASS" : "FAIL";
  const large = ratio >= LARGE_TEXT_RATIO ? "PASS" : "FAIL";
  console.log(
    `${formatRatio(ratio)} — body text ${BODY_TEXT_RATIO}:1 ${body}, large text and UI ${LARGE_TEXT_RATIO}:1 ${large}`,
  );
  process.exit(ratio >= BODY_TEXT_RATIO ? 0 : 1);
}

const args = process.argv.slice(2);
if (args[0] === "--pair" && args.length >= 3) {
  checkPair(args[1], args[2]);
} else if (args.length >= 1 && args[0] !== "--pair") {
  checkFile(args[0], args.includes("--json"));
} else {
  console.error(
    "Usage:\n  bun run check-contrast.ts <path-to-DESIGN.md> [--json]\n  bun run check-contrast.ts --pair <#RRGGBB> <#RRGGBB>",
  );
  process.exit(2);
}

File: skills/product/design-brief/scripts/check-contrast.py
#!/usr/bin/env python3
"""Check semantic DESIGN.md color pairs with WCAG 2.x contrast."""

import colorsys
import json
import math
import re
import sys
from pathlib import Path

BODY_TEXT_RATIO = 4.5

NAMED_COLORS = {
    "black": "#000000",
    "white": "#ffffff",
    "red": "#ff0000",
    "green": "#008000",
    "blue": "#0000ff",
    "yellow": "#ffff00",
    "gray": "#808080",
    "grey": "#808080",
    "orange": "#ffa500",
    "purple": "#800080",
    "rebeccapurple": "#663399",
    "transparent": "#00000000",
}


def unquote(value):
    value = value.strip()
    if len(value) >= 2 and value[0] == value[-1] and value[0] in "\"'":
        return value[1:-1]
    return value


def extract_frontmatter(markdown):
    match = re.match(r"^---\r?\n(.*?)\r?\n---", markdown, re.S)
    return match.group(1) if match else None


def parse_subtree(yaml_text, root_key):
    lines = yaml_text.splitlines()
    start = next((index for index, line in enumerate(lines) if line == f"{root_key}:"), -1)
    if start < 0:
        return {}

    root = {}
    stack = [(0, root)]
    for line in lines[start + 1 :]:
        if not line.strip() or line.lstrip().startswith("#"):
            continue
        match = re.match(r"^( +)([\w-]+):\s*(.*)$", line)
        if not match:
            break
        indent = len(match.group(1))
        while len(stack) > 1 and indent <= stack[-1][0]:
            stack.pop()
        parent = stack[-1][1]
        key = match.group(2)
        value = match.group(3)
        if value:
            parent[key] = unquote(value.split(" #", 1)[0])
        else:
            child = {}
            parent[key] = child
            stack.append((indent, child))
    return root


def parse_channel(value):
    value = value.strip()
    if value.endswith("%"):
        return max(0.0, min(255.0, float(value[:-1]) * 2.55))
    return max(0.0, min(255.0, float(value)))


def parse_alpha(value):
    value = value.strip()
    if value.endswith("%"):
        return max(0.0, min(1.0, float(value[:-1]) / 100.0))
    return max(0.0, min(1.0, float(value)))


def parse_hex(value):
    digits = value.removeprefix("#")
    if len(digits) in (3, 4):
        digits = "".join(character * 2 for character in digits)
    if len(digits) not in (6, 8) or not re.fullmatch(r"[0-9a-fA-F]+", digits):
        return None
    red, green, blue = (int(digits[index : index + 2], 16) for index in (0, 2, 4))
    alpha = int(digits[6:8], 16) / 255.0 if len(digits) == 8 else 1.0
    return red, green, blue, alpha


def parse_rgb(value):
    match = re.fullmatch(r"rgba?\((.*)\)", value, re.I)
    if not match:
        return None
    body = match.group(1).replace(",", " ")
    color_part, separator, alpha_part = body.partition("/")
    parts = color_part.split()
    if len(parts) == 4 and not separator:
        alpha_part = parts.pop()
    if len(parts) != 3:
        return None
    try:
        red, green, blue = (parse_channel(part) for part in parts)
        alpha = parse_alpha(alpha_part) if alpha_part else 1.0
    except ValueError:
        return None
    return red, green, blue, alpha


def parse_hsl(value):
    match = re.fullmatch(r"hsla?\((.*)\)", value, re.I)
    if not match:
        return None
    body = match.group(1).replace(",", " ")
    color_part, separator, alpha_part = body.partition("/")
    parts = color_part.split()
    if len(parts) == 4 and not separator:
        alpha_part = parts.pop()
    if len(parts) != 3 or not parts[1].endswith("%") or not parts[2].endswith("%"):
        return None
    try:
        hue = float(re.sub(r"(deg|turn|rad|grad)$", "", parts[0]))
        saturation = float(parts[1][:-1]) / 100.0
        lightness = float(parts[2][:-1]) / 100.0
        alpha = parse_alpha(alpha_part) if alpha_part else 1.0
    except ValueError:
        return None
    red, green, blue = colorsys.hls_to_rgb((hue % 360.0) / 360.0, lightness, saturation)
    return red * 255.0, green * 255.0, blue * 255.0, alpha


def parse_oklch(value):
    match = re.fullmatch(
        r"oklch\(\s*([\d.]+)%?\s+([\d.]+)\s+([\d.]+)(?:deg)?(?:\s*/\s*([\d.]+%?))?\s*\)",
        value,
        re.I,
    )
    if not match:
        return None
    lightness = float(match.group(1))
    if "%" in match.group(0).split()[0]:
        lightness /= 100.0
    chroma = float(match.group(2))
    hue_radians = math.radians(float(match.group(3)))
    alpha = parse_alpha(match.group(4)) if match.group(4) else 1.0
    axis_a = chroma * math.cos(hue_radians)
    axis_b = chroma * math.sin(hue_radians)
    light_l = lightness + 0.3963377774 * axis_a + 0.2158037573 * axis_b
    light_m = lightness - 0.1055613458 * axis_a - 0.0638541728 * axis_b
    light_s = lightness - 0.0894841775 * axis_a - 1.2914855480 * axis_b
    linear_red = 4.0767416621 * light_l**3 - 3.3077115913 * light_m**3 + 0.2309699292 * light_s**3
    linear_green = -1.2684380046 * light_l**3 + 2.6097574011 * light_m**3 - 0.3413193965 * light_s**3
    linear_blue = -0.0041960863 * light_l**3 - 0.7034186147 * light_m**3 + 1.7076147010 * light_s**3

    def encode(channel):
        encoded = 12.92 * channel if channel <= 0.0031308 else 1.055 * channel ** (1 / 2.4) - 0.055
        return max(0.0, min(255.0, encoded * 255.0))

    return encode(linear_red), encode(linear_green), encode(linear_blue), alpha


def parse_color(value):
    normalized = unquote(str(value)).strip().lower()
    if normalized in NAMED_COLORS:
        normalized = NAMED_COLORS[normalized]
    if normalized.startswith("#"):
        return parse_hex(normalized)
    return parse_rgb(normalized) or parse_hsl(normalized) or parse_oklch(normalized)


def composite(color, background):
    red, green, blue, alpha = color
    if alpha >= 1.0:
        return red, green, blue
    return tuple(
        foreground * alpha + backdrop * (1.0 - alpha)
        for foreground, backdrop in zip((red, green, blue), background)
    )


def luminance(color):
    def linearize(channel):
        channel /= 255.0
        return channel / 12.92 if channel <= 0.04045 else ((channel + 0.055) / 1.055) ** 2.4

    red, green, blue = color
    return 0.2126 * linearize(red) + 0.7152 * linearize(green) + 0.0722 * linearize(blue)


def contrast_ratio(first, second):
    first_luminance = luminance(first)
    second_luminance = luminance(second)
    return (max(first_luminance, second_luminance) + 0.05) / (
        min(first_luminance, second_luminance) + 0.05
    )


def resolve_reference(value, colors):
    match = re.fullmatch(r"\{colors\.([\w-]+)\}", str(value))
    return colors.get(match.group(1)) if match else value


def semantic_pairs(colors):
    pairs = []
    for token in colors:
        if token.endswith("-foreground"):
            base = token[: -len("-foreground")]
            if base in colors:
                pairs.append((base, token))
        if token.startswith("on-"):
            base = token[len("on-") :]
            if base in colors:
                pairs.append((base, token))
    if "foreground" in colors and "background" in colors:
        pairs.append(("background", "foreground"))
    secondary_text = "muted-foreground" if "muted-foreground" in colors else "on-muted" if "on-muted" in colors else None
    if secondary_text:
        for surface in ("background", "surface", "card", "muted"):
            if surface in colors and (surface, secondary_text) not in pairs:
                pairs.append((surface, secondary_text))
    return pairs


def check_pair(label, background_value, text_value):
    background = parse_color(background_value)
    text = parse_color(text_value)
    if background is None or text is None:
        return {
            "label": label,
            "status": "ERROR",
            "ratio": None,
            "message": "required CSS color could not be parsed",
        }
    if background[3] < 1.0:
        return {
            "label": label,
            "status": "ERROR",
            "ratio": None,
            "message": "translucent background has no known backdrop",
        }
    background_rgb = background[:3]
    text_rgb = composite(text, background_rgb)
    ratio = contrast_ratio(background_rgb, text_rgb)
    return {
        "label": label,
        "status": "PASS" if ratio >= BODY_TEXT_RATIO else "FAIL",
        "ratio": round(ratio, 2),
        "message": f"requires {BODY_TEXT_RATIO}:1",
    }


def run_file(path):
    try:
        markdown = Path(path).read_text(encoding="utf-8")
    except OSError as error:
        return [], f"could not read {path}: {error}"
    frontmatter = extract_frontmatter(markdown)
    if frontmatter is None:
        return [], "missing YAML frontmatter"
    colors = parse_subtree(frontmatter, "colors")
    components = parse_subtree(frontmatter, "components")
    results = []
    for background_key, text_key in semantic_pairs(colors):
        results.append(
            check_pair(
                f"colors.{text_key} on colors.{background_key}",
                colors[background_key],
                colors[text_key],
            )
        )
    for component_name, properties in components.items():
        if not isinstance(properties, dict):
            continue
        if "backgroundColor" not in properties or "textColor" not in properties:
            continue
        background_value = resolve_reference(properties["backgroundColor"], colors)
        text_value = resolve_reference(properties["textColor"], colors)
        results.append(
            check_pair(
                f"components.{component_name}",
                background_value,
                text_value,
            )
        )
    if not results:
        return [], "no checkable semantic or component color pair"
    return results, None


def main(arguments):
    if len(arguments) < 2:
        print("usage: check-contrast.py DESIGN.md [--json]", file=sys.stderr)
        return 2
    path = arguments[1]
    json_output = "--json" in arguments[2:]
    results, error = run_file(path)
    if error:
        payload = {"status": "error", "message": error, "results": []}
        print(json.dumps(payload) if json_output else f"ERROR {error}")
        return 2
    failures = [result for result in results if result["status"] != "PASS"]
    payload = {
        "status": "failed" if failures else "passed",
        "summary": {
            "checks": len(results),
            "passed": len(results) - len(failures),
            "failed": len(failures),
        },
        "results": results,
    }
    if json_output:
        print(json.dumps(payload, indent=2))
    else:
        for result in results:
            ratio = f" {result['ratio']}:1" if result["ratio"] is not None else ""
            print(f"{result['status']} {result['label']}{ratio} — {result['message']}")
    return 1 if failures else 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))
